jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/ThreadPool.java

```diff
@@ -87,4 +88,38 @@ default ThreadPoolBudget getThreadPoolBudget()
             return null;
         }
     }
+
+    /**
+     * <p>Execute a task immediately without queueing.   This may use a
+     * {@code ReservedThread}, a {@code Virtual Thread}, a call to {@link Invocable#invokeNonBlocking(Runnable)},
+     * a newly spawned thread, or direct execution.
+     *
+     * @param executor An executor that may be used
+     * @param task The task that must be executed.
+     */
+    static void mustExecute(Executor executor, Runnable task)
+    {
+        if (task == null)
+            return;
+
+        if (executor instanceof TryExecutor tryExecutor && tryExecutor.tryExecute(task))
+            return;
+
+        Executor virtual = VirtualThreads.getVirtualThreadsExecutor(executor);
+        if (virtual != null)
+            virtual.execute(task);
+
+        switch (Invocable.getInvocationType(task))
+        {
+            case NON_BLOCKING -> task.run();
+            case EITHER -> Invocable.invokeNonBlocking(task);
+            default ->
+            {
+                if (Runtime.getRuntime().freeMemory() > (10 * 1024))
```

Comment:
I don't get what this heuristic tries to achieve; the current free heap memory has nothing to do with the capacity to spawn native threads AFAICT.

Reply:
How about checking against max memory?  I know the stack does not come from the heap, but any thread will consume some heap.     

Otherwise, any other heuristic you can think of to decide between the two options below?

Reply:
@lorban another option is to try to start a thread, but if that throws an OOM, then run the task directly.   As per: https://support.oracle.com/knowledge/Middleware/462808_1.html

Reply:
That does sound a lot more reasonable. I'd be okay with such implementation.